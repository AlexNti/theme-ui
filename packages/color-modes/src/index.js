import React from 'react'
import {
  jsx,
  useThemeUI,
  merge,
  Context,
} from '@theme-ui/core'
import { get } from '@theme-ui/css'
import {
  Global,
  ThemeContext as EmotionContext
} from '@emotion/core'
import {
  toCustomProperties,
  createColorStyles,
} from './custom-properties'

const STORAGE_KEY = 'theme-ui-color-mode'
const storage = {
  get: init => window.localStorage.getItem(STORAGE_KEY) || init,
  set: value => window.localStorage.setItem(STORAGE_KEY, value),
}

// todo: update API
/*
const getMediaQuery = () => {
  const darkQuery = '(prefers-color-scheme: dark)'
  const mql = window.matchMedia ? window.matchMedia(darkQuery) : {}
  const dark = mql.media === darkQuery
  return dark && mql.matches
}
*/

export const useColorModeState = (theme = {}) => {
  const [mode, setMode] = React.useState(theme.initialColorModeName || 'default')

  // initialize state
  React.useEffect(() => {
    const stored = storage.get()
    document.body.classList.remove('theme-ui-' + stored)
    // consider prefers-color-scheme media query
    if (!stored || stored === mode) return
    setMode(stored)
  }, [])

  React.useEffect(() => {
    if (!mode) return
    storage.set(mode)
  }, [mode])

  if (process.env.NODE_ENV !== 'production') {
    if (
      theme.colors &&
      theme.colors.modes &&
      Object.keys(theme.colors.modes).indexOf(theme.initialColorModeName) > -1
    ) {
      console.warn(
        'The `initialColorModeName` value should be a unique name' +
          ' and cannot reference a key in `theme.colors.modes`.'
      )
    }
  }

  return [mode, setMode]
}

export const useColorMode = () => {
  const { colorMode, setColorMode } = useThemeUI()

  if (typeof setColorMode !== 'function') {
    throw new Error(`[useColorMode] requires the ColorModeProvider component`)
  }

  return [colorMode, setColorMode]
}

const applyColorMode = (theme, mode) => {
  if (!mode) return theme
  const modes = get(theme, 'colors.modes', {})
  // TODO: test for how this affects usage
  if (theme.useCustomProperties !== false) {
    // theme.colors = toCustomProperties(theme.colors, 'colors')
  }
  return merge.all({}, theme, {
    colors: get(modes, mode, {}),
  })
}

export const ColorMode = () =>
  jsx(Global, {
    styles: createColorStyles
  })

const BaseProvider = ({ context, children }) => {
  const theme = {...context.theme}
  if (theme.useCustomProperties !== false) {
    theme.colors = toCustomProperties(theme.colors, 'colors')
  }
  return jsx(
    EmotionContext.Provider, { value: theme },
    jsx(Context.Provider, {
      value: context,
      children
    })
  )
}

export const ColorModeProvider = ({
  children,
}) => {
  const outer = useThemeUI()
  const [colorMode, setColorMode] = useColorModeState(outer.theme)
  const theme = applyColorMode(outer.theme || {}, colorMode)
  const context = {
    ...outer,
    theme,
    colorMode,
    setColorMode,
  }

  return jsx(React.Fragment, null,
    jsx(ColorMode),
    jsx(BaseProvider, {
      context,
      children,
    })
  )
}

const noflash = `(function() { try {
  var mode = localStorage.getItem('theme-ui-color-mode');
  if (!mode) return
  document.body.classList.add('theme-ui-' + mode);
} catch (e) {} })();`

export const InitializeColorMode = () => (
  jsx('script', {
    key: 'theme-ui-no-flash',
    dangerouslySetInnerHTML: {
      __html: noflash,
    }
  })
)
