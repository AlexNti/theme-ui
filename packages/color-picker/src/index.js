/** @jsx jsx */
import { jsx } from 'theme-ui'
import { CustomPicker } from 'react-color'
import {
  Hue,
  Saturation,
  EditableInput,
} from 'react-color/lib/components/common'

const Box = ({ height = 16, ...props }) => (
  <div
    {...props}
    sx={{
      position: 'relative',
      width: 256,
      height,
    }}
  />
)

const Lens = props => <div {...props} sx={{}} />

const Handle = props => <div {...props} sx={{}} />

export default CustomPicker(props => {
  return (
    <div
      sx={{
        display: 'grid',
        gridGap: 2,
      }}>
      <Box height={256}>
        <Saturation {...props} />
      </Box>
      <Box height={8}>
        <Hue {...props} />
      </Box>
      <EditableInput
        {...props}
        value={props.hex}
        sx={{
          width: '100%',
          fontSize: 'inherit',
          fontFamily: 'inherit',
        }}
      />
    </div>
  )
})
