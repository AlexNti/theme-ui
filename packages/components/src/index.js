export { Box } from './box'
export { Flex } from './flex'
export { Button } from './button'
export { Link } from './link'
export { Text } from './text'
export { Heading } from './heading'
export { Image } from './image'
export { Card } from './card'
export { Grid } from './grid'
