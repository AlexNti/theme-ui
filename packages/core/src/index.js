// todo/figure out
// - [ ] color modes
// - [ ] custom properties
// - [ ] mdx components @theme-ui/mdx
// - [ ] better default color mode name API

// API
//  <ThemeProvider theme={theme}>
//    <ColorModeProvider>
//      <MDXProvider>
//        <ColorMode />
//        <App />
//      </MDXProvider>
//    </ColorModeProvider>
//  </ThemeProvider>

import {
  jsx as emotion,
  ThemeContext as EmotionContext,
} from '@emotion/core'
import { css, get } from '@theme-ui/css'
import {
  createContext,
  useContext,
  useState,
  useEffect,
} from 'react'
import deepmerge from 'deepmerge'
import { version as __EMOTION_VERSION__ } from '@emotion/core/package.json'

const getCSS = props => {
  if (!props.sx && !props.css) return undefined
  return theme => {
    const styles = css(props.sx)(theme)
    const raw = typeof props.css === 'function' ? props.css(theme) : props.css
    return [styles, raw]
  }
}

const parseProps = props => {
  if (!props) return null
  const next = {}
  for (let key in props) {
    if (key === 'sx') continue
    next[key] = props[key]
  }
  const css = getCSS(props)
  if (css) next.css = css
  return next
}

export const jsx = (type, props, ...children) =>
  emotion.apply(undefined, [type, parseProps(props), ...children])

export const Context = createContext({
  __EMOTION_VERSION__,
  theme: null,
})

export const useThemeUI = () => useContext(Context)

const canUseSymbol = typeof Symbol === 'function' && Symbol.for

const REACT_ELEMENT = canUseSymbol ? Symbol.for('react.element') : 0xeac7
const FORWARD_REF = canUseSymbol ? Symbol.for('react.forward_ref') : 0xeac7

const isMergeableObject = n => {
  return (
    !!n &&
    typeof n === 'object' &&
    n.$$typeof !== REACT_ELEMENT &&
    n.$$typeof !== FORWARD_REF
  )
}

const arrayMerge = (destinationArray, sourceArray, options) => sourceArray

export const merge = (a, b) =>
  deepmerge(a, b, { isMergeableObject, arrayMerge })

merge.all = (...args) => deepmerge.all(args, { isMergeableObject, arrayMerge })

const BaseProvider = ({ context, children }) =>
  jsx(
    EmotionContext.Provider, { value: context.theme },
    jsx(Context.Provider, {
      value: context,
      children
    })
  )

// possible new api
//  <ThemeProvider theme={theme}>
//  <ColorModeProvider>
//  <ComponentsProvider components={components}>

export const ThemeProvider = ({
  theme,
  children
}) => {
  const outer = useThemeUI()
  const context = {
    ...outer,
    theme: typeof theme === 'function'
      ? theme(outer.theme)
      : merge.all({}, outer.theme, theme)
  }
  return jsx(BaseProvider, {
    context,
    children
  })
}

// color modes
const STORAGE_KEY = 'theme-ui-color-mode'
const storage = {
  get: init => window.localStorage.getItem(STORAGE_KEY) || init,
  set: value => window.localStorage.setItem(STORAGE_KEY, value),
}

export const useColorState = theme => {
  const [mode, setMode] = useState(theme.initialColorModeName || 'default')

  // initialize state
  useEffect(() => {
    const stored = storage.get()
    document.body.classList.remove('theme-ui-' + stored)
    // consider prefers-color-scheme media query
    if (!stored || stored === mode) return
    setMode(stored)
  }, [])

  useEffect(() => {
    if (!mode) return
    storage.set(mode)
  }, [mode])

  if (process.env.NODE_ENV !== 'production') {
    if (
      theme.colors &&
      theme.colors.modes &&
      Object.keys(theme.colors.modes).indexOf(theme.initialColorModeName) > -1
    ) {
      console.warn(
        'The `initialColorModeName` value should be a unique name' +
          ' and cannot reference a key in `theme.colors.modes`.'
      )
    }
  }

  return [mode, setMode]
}

export const useColorMode = () => {
  const { colorMode, setColorMode } = useThemeUI()

  if (typeof setColorMode !== 'function') {
    throw new Error(`[useColorMode] requires the ColorModeProvider component`)
  }

  return [colorMode, setColorMode]
}

const applyColorMode = (theme, mode) => {
  if (!mode) return theme
  const modes = get(theme, 'colors.modes', {})
  return merge.all({}, theme, {
    colors: get(modes, mode, {}),
  })
}

export const ColorModeProvider = ({
  children,
}) => {
  const outer = useThemeUI()
  const [colorMode, setColorMode] = useColorModeState(outer.theme)
  const theme = applyColorMode(outer.theme, colorMode)
  const context = {
    ...outer,
    theme,
    colorMode,
    setColorMode,
  }

  return jsx(BaseProvider, {
    context,
    children,
  })
}

// export const ColorMode
// export const InitializeColorMode
