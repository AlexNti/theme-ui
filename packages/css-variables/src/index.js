const camelCaseToKebabCase = variableName =>
  variableName.replace(/([a-z0-9])([A-Z])/g, '$1-$2').toLowerCase()

const stringToCssVariable = (variableName, prefix) =>
  prefix
    ? `--${prefix}-${variableName}`.toLowerCase()
    : `--${variableName}`.toLowerCase()

export default (theme, prefix) => {
  const cssVariables = {}

  Object.entries(theme).forEach(([key, value]) => {
    if (Array.isArray(value)) {
      value.forEach((item, i) => {
        const variableName = stringToCssVariable(`${key}-${i}`, prefix)
        cssVariables[variableName] = item
      })
    } else if (value === Object(value)) {
      Object.entries(value).forEach(([childKey, childValue]) => {
        const variableName = stringToCssVariable(
          `${key}-${camelCaseToKebabCase(childKey)}`,
          prefix
        )
        cssVariables[variableName] = childValue
      })
    } else {
      cssVariables[key] = value
    }
  })

  return Object.entries(cssVariables)
    .map(([key, value]) => {
      return `${key}: ${value};`
    })
    .join('\n')
}
