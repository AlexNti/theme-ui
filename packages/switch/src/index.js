/** @jsx jsx */
import { jsx } from 'theme-ui'
import React, {
  useState,
  useEffect,
} from 'react'

const IconSpan = props =>
  <span
    {...props}
    css={{
      width: 24,
      height: 24,
      display: 'flex',
      alignItems: 'center',
      justifyContent: 'center',
    }}
  />

export const Thumb = props =>
  <div
    {...props}
    css={{
      width: 24,
      height: 24,
      borderRadius: 99999,
      color: 'primary',
      bg: 'background',
      border: '1px solid',
    }}
  />

export const Switch = ({
  checked,
  onClick,
  label,
  icons,
  thumb,
  ...props
}) =>
  <button
    {...props}
    type='button'
    role='switch'
    aria-checked={checked}
    aria-label={label}
    onClick={onClick}
    css={{
      appearance: 'none',
      color: 'inherit',
      bg: checked ? 'primary': 'muted',
      p: 0,
      m: 0,
      display: 'inline-flex',
      alignItems: 'center',
      width: 48,
      height: 24,
      fontFamily: 'inherit',
      fontSize: 0,
      borderRadius: 99999,
      border: 0,
      position: 'relative',
      ':focus': {
        outline: 'none',
        div: {
          boxShadow: '0 0 4px',
        }
      }
    }}>
    {icons && (
      <IconSpan>
        {icons.checked}
      </IconSpan>
    )}
    {icons && (
      <IconSpan>
        {icons.unchecked}
      </IconSpan>
    )}
    {React.cloneElement(thumb, {
      style: {
        position: 'absolute',
        top: 0,
        left: 0,
        transitionProperty: 'transform',
        transitionDuration: '0.1s',
        transitionTimingFunction: 'ease-out',
        transform: checked ? `translateX(100%)` : `translateX(0)`
      }
    })}
  </button>

Switch.defaultProps = {
  label: 'Toggle',
  icons: false,
  thumb: <Thumb />,
}

export default Switch
