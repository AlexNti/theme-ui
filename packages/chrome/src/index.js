/** @jsx jsx */
import { jsx } from 'theme-ui'
import React, { useReducer, useEffect, useRef, useState } from 'react'
import { Global } from '@emotion/core'
import { render } from 'react-dom'
import merge from 'lodash.merge'
import debounce from 'lodash.debounce'
import copyToClipboard from 'copy-to-clipboard'
import { ThemeProvider, Styled, ColorMode, useColorMode } from 'theme-ui'
import theme from './theme'
import { Colors, Modes, Typography } from '@theme-ui/editor'

const runScript = script =>
  new Promise((resolve, reject) => {
    debounce(window.chrome.devtools.inspectedWindow.eval, 100)(
      script,
      (result, err) => {
        if (err) {
          console.error(err)
          reject(err)
        }
        resolve(result)
      }
    )
  })

const mergeState = (state, next) => merge({}, state, next)

const CopyTheme = ({ theme }) => {
  const [copied, setCopied] = useState(false)
  const timer = useRef(false)

  const handleClick = () => {
    setCopied(true)
    copyToClipboard(JSON.stringify(theme))
    clearInterval(timer.current)
    timer.current = setInterval(() => setCopied(false), 1000)
  }

  return (
    <button onClick={handleClick}>{copied ? 'Copied!' : 'Copy theme'}</button>
  )
}

const Panel = ({ state, setColorMode, setTheme }) => (
  <div sx={{ p: 4, maxWidth: 400 }}>
    <header>
      <h1
        sx={{
          fontSize: 5,
          mt: 0,
          mb: 4,
        }}>
        Theme UI Devtools
      </h1>
    </header>
    <main>
      <Colors theme={state.theme} setTheme={setTheme} />
      {state.colorMode && (
        <Modes
          theme={state.theme}
          colorMode={state.colorMode}
          setColorMode={setColorMode}
        />
      )}
      <Typography theme={state.theme} setTheme={setTheme} />
    </main>
    <footer>
      <CopyTheme theme={state.theme} />
    </footer>
  </div>
)

const Editor = () => {
  theme.colorMode =
    window.chrome.devtools.panels.themeName === 'dark' ? 'dark' : 'light'

  const [state, setState] = useReducer(mergeState, {
    theme: null,
    colorMode: null,
  })

  const getTheme = () => {
    runScript(`window.__THEME_UI__.theme`).then(theme => {
      setState({ theme })
    })
  }

  const getColorMode = () => {
    runScript(`window.__THEME_UI__.colorMode`).then(colorMode => {
      setState({ colorMode })
    })
  }

  const setTheme = nextTheme => {
    const json = JSON.stringify(nextTheme)
    runScript(`window.__THEME_UI__.setTheme(${json})`)
    setState({ theme: nextTheme })
  }

  const setColorMode = nextMode => {
    runScript(`window.__THEME_UI__.setColorMode('${nextMode}')`)
  }

  useEffect(() => {
    getTheme()
    getColorMode()
  }, [])

  return (
    <ThemeProvider theme={theme}>
      <Styled.root>
        <ColorMode />
        <Global
          styles={{
            body: {
              margin: 0,
            },
          }}
        />
        {state.theme && (
          <Panel
            state={state}
            setTheme={setTheme}
            setColorMode={setColorMode}
          />
        )}
      </Styled.root>
    </ThemeProvider>
  )
}

render(<Editor />, document.getElementById('root'))
