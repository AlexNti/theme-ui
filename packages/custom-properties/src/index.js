import React, { useContext, useState, useEffect } from 'react'
import { ThemeContext, Global } from '@emotion/core'

// - [ ] OR hook into theme-ui context??

const flatten = (obj, parent, mode) => {
  const result = []
  for (const key in obj) {
    const name = [parent, key].filter(Boolean).join('.')
    const value = obj[key]

    if (key === 'styles') continue

    if (value && typeof value === 'object') {
      if (parent === 'colors.modes') {
        result.push(...flatten(value, 'colors', key))
      } else {
        result.push(...flatten(value, name))
      }
      continue
    }

    result.push({
      mode,
      name,
      value,
    })
  }
  return result
}

const toVarName = str => `--${str.replace(/\./g, '-')}`

const createStyles = theme => {
  const vars = flatten(theme)
  const styles = {
    body: {
      color: 'var(--colors-text)',
      backgroundColor: 'var(--colors-background)',
    },
  }
  vars.forEach(({ name, value, mode }) => {
    const varName = toVarName(name)
    if (mode) {
      const key = `&.${mode}`
      styles.body[key] = styles.body[key] || {}
      styles.body[key][varName] = value
    } else {
      styles.body[varName] = value
    }
  })
  return styles
}

const themeToVars = (obj, parent) => {
  const next = {}
  for (const key in obj) {
    const name = [parent, key].filter(Boolean).join('.')
    const value = obj[key]
    if (key === 'modes') continue
    if (key === 'styles') {
      next.styles = value
      continue
    }

    if (value && typeof value === 'object') {
      next[key] = themeToVars(value, name)
      continue
    }

    next[key] = `var(${toVarName(name)}, ${value})`
  }

  return next
}

const getModes = obj => {
  return [
    obj.initialColorMode,
    ...Object.keys((obj.colors && obj.colors.modes) || {}),
  ].filter(Boolean)
}

export const useTheme = () => useContext(ThemeContext)

export const CustomPropertiesProvider = ({ theme, children }) => {
  const outer = useTheme()
  const merged = { ...outer, ...theme }
  const [colorMode, setColorMode] = useState(merged.initialColorMode)
  const modes = getModes(merged)

  useEffect(() => {
    document.body.classList.remove(...modes)
    document.body.classList.add(colorMode)
  }, [colorMode])

  const styles = createStyles(merged)
  const context = themeToVars(merged)
  context.colorMode = colorMode
  context.setColorMode = setColorMode
  context.cycleColorMode = () => {
    const i = modes.indexOf(colorMode)
    const n = (i + 1) % modes.length
    setColorMode(modes[n])
  }

  return (
    <ThemeContext.Provider value={context}>
      <Global styles={styles} />
      <button
        onClick={e => {
          context.cycleColorMode()
          // setColorMode(colorMode === 'light' ? 'dark' : 'light')
        }}
      >
        {colorMode}
      </button>
      {children}
    </ThemeContext.Provider>
  )
}
