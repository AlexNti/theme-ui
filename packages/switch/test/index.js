import React from 'react'
import renderer from 'react-test-renderer'
import { Switch } from '../src'

const renderJSON = el => renderer.create(el).toJSON()

test('renders', () => {
  const json = renderJSON(
    <Switch />
  )
  expect(json).toMatchSnapshot()
})

test('renders checked', () => {
  const json = renderJSON(
    <Switch checked />
  )
  expect(json).toMatchSnapshot()
})

test('renders with icons', () => {
  const json = renderJSON(
    <Switch
      icons={{
        checked: 'on',
        unchecked: 'off',
      }}
    />
  )
  expect(json).toMatchSnapshot()
})
