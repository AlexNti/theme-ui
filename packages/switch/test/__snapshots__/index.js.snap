// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders 1`] = `
<button
  aria-label="Toggle"
  className="  css-1veo4a4"
  role="switch"
  type="button"
>
  <div
    className="css-vvwdnd"
    style={
      Object {
        "left": 0,
        "position": "absolute",
        "top": 0,
        "transform": "translateX(0)",
        "transitionDuration": "0.1s",
        "transitionProperty": "transform",
        "transitionTimingFunction": "ease-out",
      }
    }
  />
</button>
`;

exports[`renders checked 1`] = `
<button
  aria-checked={true}
  aria-label="Toggle"
  className=" active css-1lodfny"
  role="switch"
  type="button"
>
  <div
    className="css-vvwdnd"
    style={
      Object {
        "left": 0,
        "position": "absolute",
        "top": 0,
        "transform": "translateX(100%)",
        "transitionDuration": "0.1s",
        "transitionProperty": "transform",
        "transitionTimingFunction": "ease-out",
      }
    }
  />
</button>
`;

exports[`renders with icons 1`] = `
<button
  aria-label="Toggle"
  className="  css-1veo4a4"
  role="switch"
  type="button"
>
  <span
    className="css-1n11h4v"
  >
    on
  </span>
  <span
    className="css-1n11h4v"
  >
    off
  </span>
  <div
    className="css-vvwdnd"
    style={
      Object {
        "left": 0,
        "position": "absolute",
        "top": 0,
        "transform": "translateX(0)",
        "transitionDuration": "0.1s",
        "transitionProperty": "transform",
        "transitionTimingFunction": "ease-out",
      }
    }
  />
</button>
`;
